models/Version/Adapter/DatabaseVersionStorageAdapter.php

```diff
@@ -0,0 +1,161 @@
+<?php
+
+/**
+ * Pimcore
+ *
+ * This source file is available under two different licenses:
+ * - GNU General Public License version 3 (GPLv3)
+ * - Pimcore Commercial License (PCL)
+ * Full copyright and license information is available in
+ * LICENSE.md which is distributed with this source code.
+ *
+ *  @copyright  Copyright (c) Pimcore GmbH (http://www.pimcore.org)
+ *  @license    http://www.pimcore.org/license     GPLv3 and PCL
+ */
+
+namespace Pimcore\Model\Version\Adapter;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Driver\Result;
+use Doctrine\DBAL\ParameterType;
+
+class DatabaseVersionStorageAdapter implements VersionStorageAdapterInterface
+{
+    CONST versionsTableName = "versionsData";
+
+    protected function getDb() : Connection
+    {
+        return \Pimcore::getContainer()->get($this->database_connection);
+    }
+
+    public function __construct(protected string $database_connection)
+    {
+    }
+
+    public function save(int $id,
+                         int $cId,
+                         string $cType,
+                         string $storageType,
+                         string $metaData,
+                         mixed $binaryDataStream = null,
+                         string $binaryFileHash = null,
+                         int $binaryFileId = null): void
+    {
+
+        if(isset($binaryDataStream) === true &&
+            isset($binaryFileId) === false) {
+            $contents = stream_get_contents($binaryDataStream);
+        }
+
+        $this->getDb()->executeStatement("insert into " . self::versionsTableName . "(id, cid, ctype, metaData, binaryData) values (:id, :cid, :ctype, :metaData, :binaryData)",
+            [
+                'id' => $id,
+                'cid' => $cId,
+                'ctype' => $cType,
+                'metaData' => $metaData,
+                'binaryData' => $contents ?? null
+            ]);
+    }
+
+    /**
+     * @param int $id
+     * @param int $cId
+     * @param string $cType
+     * @param bool $binaryData
+     * @return mixed
+     * @throws \Doctrine\DBAL\Driver\Exception
+     * @throws \Doctrine\DBAL\Exception
+     */
+    protected function loadData(int    $id,
+                                int    $cId,
+                                string $cType,
+                                bool   $binaryData = false): mixed {
+
+        $data = null;
+        $dataColumn = $binaryData ? 'binaryData' : 'metaData';
+        $resultSet = $this->getDb()->executeQuery("SELECT " . $dataColumn . " FROM " . self::versionsTableName . " WHERE id = :id AND cid = :cid and ctype = :ctype",
```

Comment:
We should use `$this->getDb()->fetchOne()` here directly, saves some lines of code😉